### DISCUSSION TITLE
Client to client clone via ssh?

### DISCUSSION THREAD

**Discussion by c0g** [git-lfs/git-lfs#5877] [Q&A] [answered]
I'm trying to use LFS to store some data in-repo, but I need to be able push data between machines on the same network. This doesn't seem to work - git (on my mac) looks for 'git-lfs-authenticate', doesn't find it on the raspberry pi, and dies. Both mac and pi are on git lfs 3.5.1. Git on the mac is 2.45.2, pi is 2.39.2.

git lfs env on the mac:
```
(base) ~/J/armitage master » git lfs env
git-lfs/3.5.1 (GitHub; darwin arm64; go 1.23.1)
git version 2.45.2

Endpoint=https://192.168.31.2/home/tom/Repo/armitage.git/info/lfs (auth=none)
  SSH=192.168.31.2:/home/tom/Repo/armitage
LocalWorkingDir=/Users/tom/Junk/armitage
LocalGitDir=/Users/tom/Junk/armitage/.git
LocalGitStorageDir=/Users/tom/Junk/armitage/.git
LocalMediaDir=/Users/tom/Junk/armitage/.git/lfs/objects
LocalReferenceDirs=
TempDir=/Users/tom/Junk/armitage/.git/lfs/tmp
ConcurrentTransfers=8
TusTransfers=false
BasicTransfersOnly=false
SkipDownloadErrors=false
FetchRecentAlways=false
FetchRecentRefsDays=7
FetchRecentCommitsDays=0
FetchRecentRefsIncludeRemotes=true
PruneOffsetDays=3
PruneVerifyRemoteAlways=false
PruneVerifyUnreachableAlways=false
PruneRemoteName=origin
LfsStorageDir=/Users/tom/Junk/armitage/.git/lfs
AccessDownload=none
AccessUpload=none
DownloadTransfers=basic,lfs-standalone-file,ssh
UploadTransfers=basic,lfs-standalone-file,ssh
GIT_EXEC_PATH=/opt/homebrew/opt/git/libexec/git-core
git config filter.lfs.process = "git-lfs filter-process"
git config filter.lfs.smudge = "git-lfs smudge -- %f"
git config filter.lfs.clean = "git-lfs clean -- %f"
```

trying to clone:

```
(base) ~/Junk » git clone git+ssh://192.168.31.2/home/tom/Repo/armitage
Cloning into 'armitage'...
remote: Enumerating objects: 29523, done.
remote: Counting objects: 100% (29523/29523), done.
remote: Compressing objects: 100% (29459/29459), done.
remote: Total 29523 (delta 38), reused 29523 (delta 38), pack-reused 0
Receiving objects: 100% (29523/29523), 34.40 MiB | 67.35 MiB/s, done.
Resolving deltas: 100% (38/38), done.
Updating files: 100% (29417/29417), done.
<insert long wait where no one does anything here>
Downloading experiments/multimodal_output_dist/data.npz (16 MB)
Error downloading object: experiments/multimodal_output_dist/data.npz (f192af3): Smudge error: Error downloading experiments/multimodal_output_dist/data.npz (f192af3f36e702ede1de8d78b0a091f72ef6bc4f219a82079900d507fcc37b48): batch request: bash: line 1: git-lfs-authenticate: command not found: exit status 127

Errors logged to '/Users/tom/Junk/armitage/.git/lfs/logs/20241004T092817.524315.log'.
Use `git lfs logs last` to view the log.
error: external filter 'git-lfs filter-process' failed
fatal: experiments/multimodal_output_dist/data.npz: smudge filter lfs failed
warning: Clone succeeded, but checkout failed.
You can inspect what was checked out with 'git status'
and retry with 'git restore --source=HEAD :/'
[20241004T092817.524315.log](https://github.com/user-attachments/files/17255530/20241004T092817.524315.log)
```

(log attached)

I also tried ssh:// (Vs. git+ssh) and just giving the pi's address directly, no difference.

**Comment by c0g** [comment]
I cargo installed `scutiger-lfs` as in some other discussion and tried again, same errors.

I do seem able to _push_ from the pi to the Macbook, although it's going quite slowly (~ 1 MB/s when they're connected via gigabit ethernet).

**Comment by c0g** [comment, marked as answer]
Ok, (perhaps obviously) `git-lfs-transfer` needs to be available when git logs in, and by default it seem like my cargo stuff isn't available. I moved it to /usr/local/bin and its working ok now.

So the answer is:
* make sure `git-lfs-transfer` is available on the remote machine, and that `ssh remote git-lfs-transfer` works, otherwise lfs will silently fall back to git-lfs-authenticate.

**Reply by chrisd8088** [reply]
Hey, I'm glad you found a solution.  You're absolutely correct that `git-lfs-transfer` is required to support Git LFS batch requests and object transfers entirely over SSH; otherwise, the client falls back to using `git-lfs-authenticate` and then making object transfers over HTTP.

**Reply by c0g** [reply]
Is it possible or planned to have this installed by the default installer? It took me a little while to work out what was going wrong.

**Reply by chrisd8088** [reply]
In general, the choice of what to include in an upstream package—like the Git for Windows [project](https://gitforwindows.org/), or the Homebrew `git-lfs` [formula](https://formulae.brew.sh/formula/git-lfs), or the packages maintained by various Linux distributions, such as Debian's `git-lfs` [package](https://packages.debian.org/sid/git-lfs)—is (mostly) outside our control.

However, we do build assets for each of our [releases](https://github.com/git-lfs/git-lfs/releases), and some Linux packages which we [distribute](https://packagecloud.io/app/github/git-lfs/search) on Packagecloud.

But we don't include other software in these packages, such as the `git-lfs/lfs-test-server` [program](https://github.com/git-lfs/lfs-test-server), which can be used for testing the original HTTP-based object transfer protocol for Git LFS.

The rationale for that, among other things, is that we expect most users only want the Git LFS client, and not any server-side implementations of the HTTP or SSH object transfer protocols.

---

One other point to consider is that at the moment we don't have an official `git-lfs/git-lfs-transfer` project with an implementation of the server-side part of the SSH transfer protocol for Git LFS.

There is, of course, the Rust program maintained by the developer of the SSH transfer protocol for Git LFS, in the `bk2204/scutiger` [project](https://github.com/bk2204/scutiger/tree/614a4f01cfbe6aafa355ca92ed6d1581b5cfc088/scutiger-lfs#git-lfs-transfer), and we use that in our CI test suite.  That project's documentation states:

> `git-lfs-transfer` provides an experimental server-side implementation of the proposed Git LFS pure SSH-based protocol.

There are also other non-experimental implementations now as well, in GitLab and, I believe, Gitea.  Plus other efforts like the `charmbracelet/git-lfs-transfer` [project](https://github.com/charmbracelet/git-lfs-transfer).

But if we were going to provide releases of a server implementation of the SSH object transfer protocol for Git LFS, I think we'd want to start by developing an official `git-lfs/git-lfs-transfer` project in the first place, which would take some time.

**Reply by c0g** [reply]
Gotcha, thanks for the response.
